src/noding/MCIndexSegmentSetMutualIntersector.cpp

```diff
@@ -124,14 +100,22 @@ MCIndexSegmentSetMutualIntersector::setBaseSegments(SegmentString::ConstVect* se
 void
 MCIndexSegmentSetMutualIntersector::process(SegmentString::ConstVect* segStrings)
 {
-    processCounter = indexCounter + 1;
-    nOverlaps = 0;
+    if (!indexBuilt) {
+        for (auto& mc: indexChains) {
+            mc.setId(indexCounter++);
```

Comment:
One thing I was trying to figure out is why we need to set these id variables. According to the comments it's so we can avoid comparing the same pair of chains twice, or comparing a chain itself. If that's all it's for, we should just be able to use the chain addresses to avoid the same thing, but doing this causes a test failure (I didn't look into the specifics but was hoping it was just a component ordering type thing)

Reply:
The motivation for getting rid of `id` was that `MonotoneChain` takes up 72 bytes and if that can be brought down to 64 if seems like you'd get faster access to the vector of chains

Reply:
The `id` is [used](https://github.com/libgeos/geos/blob/master/src/noding/MCIndexNoder.cpp#L85) in `MCIndexNoder`. It ensures that each two chains are compared once only.

Reply:
Not sure why it's being referenced in this class, though.  It isn't in JTS.  Perhaps vestigial remains of something that was subsequently removed in JTS?

Reply:
Yes, but you can ensure the same thing in C++ with an address comparison

Reply:
Gotcha.  Seems tacky somehow (but no worse than Shewchuk's Triangle implementation, where he used the low-order address bits as an index into a 3-element array!)

Reply:
I think the word you were looking for is "elegant" :)